File: leetcode/202.py
"""
Runtime: 32 ms, faster than 80.74% of Python3 online submissions for Happy Number.
Memory Usage: 14.2 MB, less than 46.84% of Python3 online submissions for Happy Number.
"""


class Solution:
    def isHappy(self, n: int) -> bool:
        def nxt(m):
            ans = 0
            while m:
                r = m % 10
                m //= 10
                ans += r ** 2

            return ans

        n1, n2 = n, n
        while n1 != 1 and n2 != 1:
            n1 = nxt(n1)
            n2 = nxt(nxt(n2))
            if n1 == n2 and n1 != 1 and n2 != 1:
                return False

        return True

File: leetcode/202.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Happy Number.
// Memory Usage: 5.9 MB, less than 62.15% of C++ online submissions for Happy Number.
class Solution {
 public:
  int NextHappyNumber(int x, int n) {
    // push x to the next happy number, n times
    while (n--) {
      int sum_of_squares = 0;
      while (x) {
        sum_of_squares += pow((x % 10), 2);
        x /= 10;
      }
      x = sum_of_squares;
    }

    return x;
  }

  bool isHappy(int n) {
    int x = n, y = n;

    // this will work only because we were told
    // there's a cycle, or it ends in 1
    do {
      x = NextHappyNumber(x, 1);
      y = NextHappyNumber(y, 2);
      if (x == y && x != 1) {
        return false;
      }
    } while (x != y);
    return true;
  }
};
